ctapipe_io_lst/calibration.py

```diff
@@ -0,0 +1,700 @@
+
+    r1_sample_start = IntTelescopeParameter(
+        default_value=3,
+        help='Start sample for r1 waveform',
+        allow_none=True,
+    ).tag(config=True)
+
+    r1_sample_end = IntTelescopeParameter(
+        default_value=39,
+        help='End sample for r1 waveform',
+        allow_none=True,
+    ).tag(config=True)
+
+    drs4_pedestal_path = TelescopeParameter(
+        trait=Path(exists=True, directory_ok=False),
+        allow_none=True,
+        default_value=None,
+        help='Path to the LST pedestal file',
+    ).tag(config=True)
+
+    calibration_path = Path(
+        exists=True, directory_ok=False,
+        help='Path to LST calibration file',
+    ).tag(config=True)
+
+    drs4_time_calibration_path = TelescopeParameter(
+        trait=Path(exists=True, directory_ok=False),
+        help='Path to the time calibration file',
+        default_value=None,
+        allow_none=True,
+    ).tag(config=True)
+
+    calib_scale_high_gain = FloatTelescopeParameter(
+        default_value=1.18,
+        help='High gain waveform is multiplied by this number'
+    ).tag(config=True)
+
+    calib_scale_low_gain = FloatTelescopeParameter(
+        default_value=1.09,
+        help='Low gain waveform is multiplied by this number'
+    ).tag(config=True)
+
+    gain_selector_type = create_class_enum_trait(
+        GainSelector,
+        default_value='ThresholdGainSelector',
+    )
+
+    def __init__(self, subarray, config=None, parent=None, **kwargs):
+        """
+        The R0 calibrator for LST data.
+        Fill the r1 container.
+
+        Parameters
+        ----------
+        """
+        super().__init__(
+            subarray=subarray, config=config, parent=parent, **kwargs
+        )
+
+        self.mon_data = None
+        self.last_readout_time = {}
+        self.first_cap = {}
+        self.first_cap_old = {}
+        self.fbn = {}
+        self.fan = {}
+
+        for tel_id in self.subarray.tel:
+            shape = (N_GAINS, N_PIXELS, N_CAPACITORS_PIXEL)
+            self.last_readout_time[tel_id] = np.zeros(shape)
+
+            shape = (N_GAINS, N_PIXELS)
+            self.first_cap[tel_id] = np.zeros(shape, dtype=int)
+            self.first_cap_old[tel_id] = np.zeros(shape, dtype=int)
+
+        # set the right default for our default selector, change back afterwards
+        # to not impact other code.
+        old_default = ThresholdGainSelector.threshold.default_value
+        ThresholdGainSelector.threshold.default_value = 20.9
+        self.gain_selector = GainSelector.from_name(
+            self.gain_selector_type, parent=self
+        )
+
+        # to make sure it is accessed and the default is set before setting back
+        if isinstance(self.gain_selector, ThresholdGainSelector):
+            self.log.debug(
+                f'Using gain selection threshold: {self.gain_selector.threshold}'
+            )
+        ThresholdGainSelector.threshold.default_value = old_default
```

Comment:
Hi @maxnoe, 
I do not understand this part of the code, why do you make this game with the default value?
From where did you get the 20.9 value and why the threshold  is not defined by a trailet?

Reply:
a) It is defined via a traitlet. The `threshold` is a traitlet of the `ThresholdGainSelector` and can be set via the config system.

b) I do the default game because I want to have the gain selector here to have a different default than "default default" in ctapipe, which is 4000. Since we are doing the gain selection after converting to pe, 4000 is not a good default.

c) I got the 20.9 from lstchain, and the images are matched well with lstchain so it seems to be the right one. Remeber, it is in terms of pe, since we do the calibration steps before and only then choose gain.

Reply:
> a) It is defined via a traitlet. The `threshold` is a traitlet of the `ThresholdGainSelector` and can be set via the config system.
> b) I do the default game because I want to have the gain selector here to have a different default than "default default" in ctapipe, which is 4000. Since we are doing the gain selection after converting to pe, 4000 is not a good default.

yes, you put the selector as trailet not the threshold (by the way, in the EvB will have a threshold selector, so it is not so clear it is necessary to have it as a trailet).
If you put the threshold as trailet we can avoid the "default game" and things are more clear

> c) I got the 20.9 from lstchain, and the images are matched well with lstchain so it seems to be the right one. Remember, it is in terms of pe, since we do the calibration steps before and only then choose gain.

Yes I know, from where did you get this value in lstchain? The present threshold is 3500 ADC which corresponds approximately to 3500/80 = 43 pe.  For example, the peak of the standard FF event is 2700 ADC/ ~30 pe ...

Reply:
From a config file where thomas replaced it. Ok, I can change the threshold to 43 pe.

Reply:
FYI the value originally comes from here:
https://indico.cta-observatory.org/event/2475/contributions/21864/attachments/16423/21354/20191104_LST_vuillaume.pdf
Since it's more than 1 year old now, it could be that it's not adapted.

Also, a threshold in p.e. is less "precise" than in ADC due to pixel to pixel variations right? Would 43 p.e. be conservative enough?

Reply:
The pulse height of a single photoelectron is 25ADC. What we got from F-factor method (80ADC*ns/pe) is an integrated charge.
This threshold (3500ADC) will be applied to the pulse height, right?
If so, 3500ADC corresponds to 140 pe.
(We have designed a high gain channel with a better bandwidth to cover below 100-200 pe.)

Reply:
the 20.9 threshold was corresponding to phe/ns for the integrated signal, right @vuillaut ?

Reply:
Yes exactly, it's a value in p.e. **per sample** (or `/ns` as you write).

Reply:
As is needed since the gain selector does `any(waveform > threshold)`

Reply:
To be more concrete, I attach the HG waveform in ADC and the gain selected waveform in pe. In the case of run 2323 the max is at ~60 pe and the present threshold at 3500 ADC selects LG values (indeed the HG waveform is often saturated). In the case of Run 2320 the max is around 3000 ADC, which corresponds to ~35 pe. So a threshold at 40-45 pe seems to me reasonable. But we can obviously perform more details studies 
![waveform_run_2323 0000](https://user-images.githubusercontent.com/32127708/100076091-18dfbc80-2e41-11eb-9940-423a01cf25f5.png)
![waveform_run_2320 0000](https://user-images.githubusercontent.com/32127708/100076113-1f6e3400-2e41-11eb-909a-0bd8bf0bc90c.png)

Reply:
Concerning the value of 25ADC/pe, this value was given from a single photoelectron waveform, with FWHM of 3ns.
This relation (60pe vs 3000ADC) can be applied only to calibration laser event which has a wider pulse width. So I think we need a careful study for Cherenkov events as Franca suggested.

Reply:
The ADC to pe conversion that we apply to laser events (in the above plots) are the same that we apply to cosmic events, so concerning the peak value, there should be no difference. But indeed we should make a fast study with cosmics to clearly document the criteria for the choice of the threshold.

Reply:
So to clarify the value should be increased from the current 20.9 to 40?

Reply:
Something like that, but let me to make some tests to be more precise. Let you know

Reply:
Hi, 
here a fast cross-check for the threshold value in pe: in the plot the colour correspond to the present HG and LG selection, based on the threshold at 3500 ADC. The y axis shows the HG waveform maximum in pe. I  would say that a threshold at 40 pe well correspond to the present selection, with a reasonable margin from the saturation value (given that the effective threshold will depend on the scaling factors)

![HG_Max_vs_LG_charge_run_3004](https://user-images.githubusercontent.com/32127708/100468474-c6b6c980-30d4-11eb-8cc1-63fa650f28ce.png)

Reply:
THe scaling is applied later in ctapipe, so it does not affect gain selection currently.

Reply:
yes, but, generally, one could have a calibration move over time (e.g, hopefully part of the present scaling one day will be included in the calibration coefficients).

By the way, I got the confirmation of Julien Houles that the EvB will perform the gain selection at ADC level. Personally, I would implement here what we have at present in lstchain and what we expect to have online in a short future. I do not really understand this parenthesis with threshold based on pe level (the change is  just to move the selection four lines above). Perhaps one day @moralejo will explain it to me ;-)

Reply:
Ok, yes, will do it in adc counts then. 3500 was it, right?

Reply:
yes

Reply:
> yes, but, generally, one could have a calibration move over time (e.g, hopefully part of the present scaling one day will be included in the calibration coefficients).
> 
> By the way, I got the confirmation of Julien Houles that the EvB will perform the gain selection at ADC level. Personally, I would implement here what we have at present in lstchain and what we expect to have online in a short future. I do not really understand this parenthesis with threshold based on pe level (the change is just to move the selection four lines above). Perhaps one day @moralejo will explain it to me ;-)

I have no strong opinion, really. I the last discussion I said I preferred the p.e. because of spikes, but as @maxnoe pointed out, one can do it in counts after the spike removal and before the conversion to pe.

Reply:
as in lstchain ;-)

Reply:
I moved it infront of the pe conversion and changed the default to 3500